cpp/include/cudf/aggregation.hpp

```diff
@@ -599,6 +609,245 @@ std::unique_ptr<Base> make_udf_aggregation(udf_type type,
+      {
+        auto const& value = attr.value;
+        auto const hash_value =
+          std::visit(cudf::detail::visitor_overload{
+                       [](auto const& val) { return std::hash<int>{}(static_cast<int>(val)); },
+                       [](std::unique_ptr<aggregation> const& val) { return val->do_hash(); }},
+                     value);
+        return std::hash<std::size_t>{}(value.index()) ^ hash_value;
+      }
+    };  // struct hash
+
+    /**
+     * @brief Equality comparison functor for `data_attribute`.
+     */
+    struct equal_to {
+      /**
+       * @brief Check if two data attributes are equal.
+       * @param lhs The left-hand side data attribute
+       * @param rhs The right-hand side data attribute
+       * @return True if the two data attributes are equal
+       */
+      bool operator()(data_attribute const& lhs, data_attribute const& rhs) const
+      {
+        auto const& lhs_val = lhs.value;
+        auto const& rhs_val = rhs.value;
+        if (lhs_val.index() != rhs_val.index()) { return false; }
+        return std::visit(cudf::detail::visitor_overload{
+                            [](auto const& lhs_val, auto const& rhs_val) {
+                              if constexpr (std::is_same_v<decltype(lhs_val), decltype(rhs_val)>) {
+                                return lhs_val == rhs_val;
+                              } else {
+                                return false;
+                              }
+                            },
+                            [](std::unique_ptr<aggregation> const& lhs_val,
+                               std::unique_ptr<aggregation> const& rhs_val) {
+                              return lhs_val->is_equal(*rhs_val);
+                            }},
+                          lhs_val,
+                          rhs_val);
+      }
+    };  // struct equal_to
+  };    // struct data_attribute
+
+  /**
+   * @brief Set of attributes for the input data that is needed for computing the aggregation.
+   */
+  using data_attributes_set_t =
+    std::unordered_set<data_attribute, data_attribute::hash, data_attribute::equal_to>;
+
+  /**
+   * @brief Return a set of attributes for the data that is needed for computing the aggregation.
+   *
+   * If this function is not overridden, an empty set is returned. That means all the data
+   * attributes (except results from other aggregations in groupby) will be needed.
+   *
+   * @return A set of `data_attribute`
+   */
+  [[nodiscard]] virtual data_attributes_set_t get_required_data() const { return {}; }
+
+  /**
+   * @brief Hold all possible types of the data that is passed to the derived class for executing
+   * the aggregation.
+   */
+  using input_data_t = std::variant<column_view, size_type, device_span<size_type const>>;
+
+  /**
+   * @brief Input to the aggregation, mapping from each data attribute to its actual data.
+   */
+  using host_udf_input = std::
```

Comment:
```suggestion
  using input_map_t = std::
```
nit: using `_t` suffix for consistency